File: LuaOpenResty/app/repositories/task_repository.lua
local DB = require("db.db")
local Task = require("app.models.task")

local _M = {}

function _M.create(user_id, title, description)
    local db = DB.connect()
    local res, err = db:execute("INSERT INTO tasks (user_id, title, description, done) VALUES (?, ?, ?, 0)", {user_id, title, description})
    db:close()

    if not res then return nil, err end
    return Task.new(res.last_insert_id, user_id, title, description, false), nil
end

function _M.find_by_user(user_id)
    local db = DB.connect()
    local rows, err = db:query("SELECT * FROM tasks WHERE user_id = ?", {user_id})
    db:close()

    if not rows then return {}, err end

    local tasks = {}
    for _, row in ipairs(rows) do
        table.insert(tasks, Task.new(row.id, row.user_id, row.title, row.description, row.done))
    end
    return tasks
end

function _M.find_by_id(id)
    local db = DB.connect()
    local rows, err = db:query("SELECT * FROM tasks WHERE id = ?", {id})
    db:close()

    if not rows or #rows == 0 then return nil, err end
    local row = rows[1]
    return Task.new(row.id, row.user_id, row.title, row.description, row.done)
end

function _M.update(id, title, description)
    local db = DB.connect()
    local res, err = db:execute("UPDATE tasks SET title = ?, description = ? WHERE id = ?", {title, description, id})
    db:close()

    if not res then return nil, err end
    if res.changes == 0 then return nil, "not found" end
    return true
end

function _M.mark_done(id)
    local db = DB.connect()
    local res, err = db:execute("UPDATE tasks SET done = 1 WHERE id = ?", {id})
    db:close()

    if not res then return nil, err end
    if res.changes == 0 then return nil, "not found" end
    return true
end

function _M.delete(id)
    local db = DB.connect()
    local res, err = db:execute("DELETE FROM tasks WHERE id = ?", {id})
    db:close()

    if not res then return nil, err end
    return true
end

return _M

File: Cpp/src/repositories/TaskRepository.cc
#include "TaskRepository.h"
#include "../db/Database.h"
#include "../db/StmtGuard.h"
#include <iostream>

namespace repositories {

    int TaskRepository::create(const models::Task& task) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "INSERT INTO tasks (title, description, is_done, user_id) VALUES (?, ?, ?, ?);";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_text(stmt, 1, task.title.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 2, task.description.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_int(stmt, 3, task.is_done ? 1 : 0);
        sqlite3_bind_int(stmt, 4, task.user_id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }

        return (int)sqlite3_last_insert_rowid(db);
    }

    std::optional<models::Task> TaskRepository::findById(int id) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "SELECT id, title, description, is_done, user_id FROM tasks WHERE id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, id);

        if (sqlite3_step(stmt) == SQLITE_ROW) {
            models::Task task;
            task.id = sqlite3_column_int(stmt, 0);
            task.title = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            task.description = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
            task.is_done = sqlite3_column_int(stmt, 3) != 0;
            task.user_id = sqlite3_column_int(stmt, 4);
            return task;
        }

        return std::nullopt;
    }

    std::vector<models::Task> TaskRepository::findAllByUserId(int userId) {
        std::vector<models::Task> tasks;
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "SELECT id, title, description, is_done, user_id FROM tasks WHERE user_id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, userId);

        while (sqlite3_step(stmt) == SQLITE_ROW) {
            models::Task task;
            task.id = sqlite3_column_int(stmt, 0);
            task.title = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            task.description = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
            task.is_done = sqlite3_column_int(stmt, 3) != 0;
            task.user_id = sqlite3_column_int(stmt, 4);
            tasks.push_back(task);
        }

        return tasks;
    }

    void TaskRepository::update(const models::Task& task) {
        sqlite3* db = db::Database::getInstance().getDb();

        std::string sql = "UPDATE tasks SET ";
        bool first = true;
        bool hasUpdate = false;

        if (!task.title.empty()) {
            sql += "title = ?";
            first = false;
            hasUpdate = true;
        }
        if (!task.description.empty()) {
            if (!first) sql += ", ";
            sql += "description = ?";
            hasUpdate = true;
        }

        if (!hasUpdate) return;

        sql += " WHERE id = ?;";

        sqlite3_stmt* stmt;
        if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        int index = 1;
        if (!task.title.empty()) {
            sqlite3_bind_text(stmt, index++, task.title.c_str(), -1, SQLITE_STATIC);
        }
        if (!task.description.empty()) {
            sqlite3_bind_text(stmt, index++, task.description.c_str(), -1, SQLITE_STATIC);
        }
        sqlite3_bind_int(stmt, index, task.id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }
    }

    void TaskRepository::markAsDone(int id) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "UPDATE tasks SET is_done = 1 WHERE id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }
    }

    void TaskRepository::deleteById(int id) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "DELETE FROM tasks WHERE id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }
    }

    bool TaskRepository::exists(int id) {
        return findById(id).has_value();
    }
}
